### PR TITLE
Sprint 5

### PR DESCRIPTION
**Pull request by bbusagroningen**
In this print we have added the following features:

* Undo button
* Redo button
* Tests for multiple factories

### CODE REVIEW

File: Switchboard/docs/config/processors/docLinks.js [hunk 1]
```diff
@@ -106,31 +106,47 @@ module.exports = function docLinksProccesor(templateEngine) {
...
 
             };
-
+            
+            
 
             var parseTypeApplication = function (type) {
                 if (type.expression.name === 'Array') {
-                    return parseName('Array') + '.<' + parseType(type.applications[0]) + '>';
+
```
Comment (lauraBaakman) [inline Switchboard/docs/config/processors/docLinks.js]: Multiple levels of abstraction; if you need a comment to explain what is happening in the bodies of your if-else you should have used functions.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
+var LinkedFieldsClass = (function (
+        $uibModal,
+        ResourcesFactory,
+        ConfirmModalFactory,
+        UndoManagerFactory) {
+
+    var linkedFieldsResource = ResourcesFactory.getResource("linkedFields");
+
+    var Tick = "\u2714";
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: 👍

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        var removeElement = function (elementToRemove) {
+
+            var currentIndex = linkedFields.indexOf(elementToRemove);
+
+            undoManager.add({
+                undo: function () {
+
+                    if (elementToRemove.id === undefined) {
+
+                        // The element has been removed.
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Multiple levels of abstraction; if you need a comment to explain what is happening in the bodies of your if-else you should have used functions.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        var linkedFields = [];
+
+        var removeElement = function (elementToRemove) {
+
+            var currentIndex = linkedFields.indexOf(elementToRemove);
+
+            undoManager.add({
+                undo: function () {
+
+                    if (elementToRemove.id === undefined) {
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: You might want to replace ``elementToRemove.id === undefined`` with something like ``hasBeenPushedToBackEnd(elementToRemove)``, I think that would really help with the readability.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                        // Add into the old position 
+                        linkedFields.insert(currentIndex, elementToRemove);
+                    }
+                },
+                redo: function () {
+
+                    // Get the new current position as it might have changed
+                    currentIndex = linkedFields.indexOf(elementToRemove);
+
+                    if (elementToRemove.id === undefined) {
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: ``hasBeenPushedToBackEnd(elementToRemove)``same as above.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+                        // Is not yet in back-end only need to remove it from
+                        // the items to add
+                        addStack.deleteByValue(elementToRemove);
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Multiple levels of abstraction; if you need a comment to explain what is happening in the bodies of your if-else you should have used functions.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+                        // Has not yet been removed from back end
+                        removeStack.push(elementToRemove);
+                    }
+
+                    linkedFields.deleteByIndex(currentIndex);
+
+                }
+            });
+
+            if (elementToRemove.id === undefined) {
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: You have a lot of vertical white space in here, it makes the code harder to read for me, but if it works for you...

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+         * @name LinkedFieldsFactory#addLinkedField
+         * @description
+         * Gets the information of the connection
+         * 
+         * @param {connections/LinkedFieldsFactory#LinkedFields.LinkedField} 
+         * newLinkedField The new field to add 
+         */
+        this.addLinkedField = function (newLinkedField) {
+
+            // TODO: Id need to be added when respond from back end
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Please link the Trello card with this todo.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+            newField.close = function () {
+                closeElement(newField);
+            };
+
+            var currentIndex;
+
+            undoManager.add({
+                undo: function () {
+
+                    // Update curent index as t might have changed depending on
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Multiple levels of abstraction; if you need a comment to explain what is happening in a chunk of code you should have used functions.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+
+                        addStack.deleteByValue(newField);
+                        
+                    } else {
+                        removeStack.push(newField);
+                    }
+                },
+                redo: function () {
+
+                    if (newField.id === undefined) {
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Multiple levels of abstraction; if you need a comment to explain what is happening in a chunk of code you should have used functions.

Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: ``hasBeenPushedToBackEnd(newField)``?

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+         * @returns {Boolean} Returns True if it has unsaved changes
+         */
+        this.getChangeText = function () {
+            if (this.hasChanges()) {
+                return changesMade;
+            }
+            return changesSaved;
+        };
+
+        var fakeIndex = 0;
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: What is this ``fakeIndex``?

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+         * @name LinkedFieldsFactory#saveChanges
+         * @description
+         * Saves the changes with the backend
+         */
+        this.saveChanges = function () {
+
+            addStack.forEach(function (elementToAdd) {
+                elementToAdd.id = "FAKE_" + fakeIndex++;
+
+                // Need to be send to back end
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Is this a todo? If so please link the trello card as a comment.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+            });
+
+            removeStack.forEach(function (elementToRemove) {
+                delete elementToRemove.id;
+
+                // Need to be send to back end
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Is this a todo? If so please link the trello card as a comment.

File: Switchboard/public/app/components/connections/NewLinkedAPIModalController.js [hunk 14]
```diff
@@ -10,32 +14,102 @@ define(["shared/Registry", "shared/api/APIFactory"], function (Registry) {
...
         $scope.submit = function () {
-            if ($scope.select.source === undefined
-                    || $scope.select.destination === undefined
-                    || !$scope.select.name) {
-                alert("Some of the fields are empty!");
+
+            var emptyFound = false;
+
+            angular.forEach($scope.select, function (field) {
+                if (!emptyFound && (field === undefined || field === "")) {
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/NewLinkedAPIModalController.js]: What are you testing in the guard of this if?

File: Switchboard/public/app/components/connections/NewLinkedFieldModalController.js [hunk 15]
```diff
@@ -9,37 +13,59 @@ define(["shared/Registry"], function (Registry) {
...
+            actionFieldId: undefined
         };
 
+        var connection = new ConnectionFactory(connectionId);
+        $scope.connection = connection;
+
         $scope.submit = function () {
-            if ($scope.select.source === undefined
-                    || $scope.select.destination === undefined) {
+            if ($scope.select.hookFieldId === undefined
```
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/NewLinkedFieldModalController.js]: What are you testing in this guard?

File: Switchboard/public/app/defaults.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
+define([], function () {
+
+    var REMOVE_NO_ELEMENTS = 0;
```
Comment (lauraBaakman) [inline Switchboard/public/app/defaults.js]: 👍

**Comment by lauraBaakman** [comment]
For each comment that you resolve please add the link to the commit that resolved it.

File: Switchboard/public/app/components/connections/NewLinkedFieldModalController.js (hunk 15 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/NewLinkedFieldModalController.js]: Checking whether or not, the selected fields are empty. If they are empty, then there will be an error message.

File: Switchboard/public/app/components/connections/NewLinkedAPIModalController.js (hunk 14 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/NewLinkedAPIModalController.js]: Checking whether or not, the selected fields are empty. If they are empty, then there will be an error message. Should I comment this section?

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 11 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: this is needed so esLint wouldn't have an issue with magic number. There is this option or we could use eslint ignore. Which one would you prefer?

Comment (killje) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: This is for currently adding an unique index to added fields while its not yet been pushed to the back end and getting a real id.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 8 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: https://trello.com/c/6UrkgN3V

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 13 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: https://trello.com/c/2WoJKI6t

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 12 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: https://trello.com/c/2WoJKI6t

**Comment by bbusagroningen** [comment]
Travis fails, because blueprint is not up to date. All other checks are fine.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 9 above)
Comment (bbusagroningen) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: MLA fixed, comments are added for further clearance. But now, they are not required, but I have decided to keep them there.

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 11 above)
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: I would have used ``frontEndIndex``, but if this works for you 👍 .

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 9 above)
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: I am no fan of keeping the comments, as they still rot, but if that's what you really want, go for it. 👍

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 8 above)
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: 👍

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 9 above)
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: Same as before 👍

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 12 above)
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: 👍

File: Switchboard/public/app/components/connections/LinkedFieldsFactory.js (hunk 13 above)
Comment (lauraBaakman) [inline Switchboard/public/app/components/connections/LinkedFieldsFactory.js]: 👍

**Comment by lauraBaakman** [comment]
@bbusagroningen don't forget the tagging and stuff. @Carolien989 @killje @bbusagroningen don't forget to rebase.

**Pull request merged**